cpp/src/arrow/compute/row/grouper_benchmark.cc

```diff
@@ -0,0 +1,122 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/util/key_value_metadata.h"
+#include "arrow/util/string.h"
+#include "benchmark/benchmark.h"
+
+#include "arrow/compute/row/grouper.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/testing/random.h"
+#include "arrow/util/benchmark_util.h"
+
+namespace arrow {
+namespace compute {
+
+constexpr auto kSeed = 0x0ff1ce;
+
+static ExecBatch MakeRandomExecBatch(const DataTypeVector& types, int64_t num_rows,
+                                     double null_probability,
+                                     int64_t alignment = kDefaultBufferAlignment,
+                                     MemoryPool* memory_pool = nullptr) {
+  random::RandomArrayGenerator rng(kSeed);
+  auto num_types = static_cast<int>(types.size());
+
+  // clang-format off
+  auto metadata = key_value_metadata(
+      {
+        "null_probability",
+        "true_probability",
+        "unique"
+      },
+      {
+          internal::ToChars(null_probability),
+          internal::ToChars(null_probability),                     // for boolean type
+          internal::ToChars(static_cast<int32_t>(num_rows * 0.5))  // for string type
+      });
+  // clang-format on
+
+  std::vector<Datum> values;
+  values.resize(num_types);
+  for (int i = 0; i < num_types; ++i) {
+    auto field = ::arrow::field("", types[i], metadata);
+    values[i] = rng.ArrayOf(*field, num_rows, alignment, memory_pool);
+  }
+
+  return ExecBatch(std::move(values), num_rows);
+}
+
+static void GrouperBenchmark(benchmark::State& state, const ExecSpan& span,
+                             const std::unique_ptr<Grouper>& grouper) {
+  for (auto _ : state) {
+    ASSIGN_OR_ABORT(auto group_ids, grouper->Consume(span));
+  }
```

Comment:
Is the `Grouper` stateless?  Would it make more sense to do this in batches?

```
for (auto _ : state) {
  // Create new grouper
  grouper = ...
  // 256 is just some arbitrary batch size to make sure initialization of the grouper doesn't dominate the benchmark
  for (int i = 0; i < 256; i++) {
    ASSIGN_OR_ABORT(auto group_ids, grouper->Consume(span));
  }
}
```

Or would it maybe make sense to do one run first to prime the result:

```
// Prime the grouper
ASSIGN_OR_ABORT(auto init, grouper->Consume(span));
for (auto _ : state) {
  ...
```

This is only a minor nit, feel free to ignore.  In the grand scheme of things, if we are getting enough iterations, this probably doesn't really matter.

Reply:
Agree with this!